sdk/src/main/java/com/kin/ecosystem/history/presenter/OrderHistoryPresenter.kt

```diff
@@ -0,0 +1,282 @@
+                            private val eventLogger: EventLogger) : BasePresenter<IOrderHistoryView>(), IOrderHistoryPresenter {
+
+    private var earnOrderList: MutableList<Order> = ArrayList()
+    private var spendOrderList: MutableList<Order> = ArrayList()
+
+    private var balanceObserver: Observer<Balance>? = null
+    private var currentBalance: Balance = blockchainSource.balance
+    private var publicAddress: String = blockchainSource.publicAddress
+
+    private var currentPendingOrder: Order? = null
+    private var completedOrderObserver: Observer<Order>? = null
+
+    override fun onAttach(view: IOrderHistoryView) {
+        super.onAttach(view)
+        eventLogger.send(OrderHistoryPageViewed.create())
+        getOrderHistoryList()
+        listenToOrders()
+    }
+
+    override fun onResume() {
+        updateMenuSettingsIcon()
+    }
+
+    override fun onPause() {
+        removeBalanceObserver()
+    }
+
+    override fun onDetach() {
+        super.onDetach()
+        completedOrderObserver?.let {
+            orderRepository.removeOrderObserver(it)
+            completedOrderObserver = null
+        }
+    }
+
+    private fun getOrderHistoryList() {
+        val cachedOrderListObj = orderRepository.allCachedOrderHistory
+        cachedOrderListObj?.let {
+            syncNewOrders(it)
+        }
+        view?.setEarnList(earnOrderList)
+        view?.setSpendList(spendOrderList)
+        orderRepository.getAllOrderHistory(object : KinCallback<OrderList> {
+            override fun onResponse(orderHistoryList: OrderList) {
+                syncNewOrders(orderHistoryList)
+            }
+
+            override fun onFailure(exception: KinEcosystemException) {
+
+            }
+        })
+    }
+
+    private fun syncNewOrders(newOrdersListObj: OrderList) {
+        val newEarnList: MutableList<Order> = ArrayList()
+        val newSpendList: MutableList<Order> = ArrayList()
+        splitByType(newOrdersListObj.orders, newEarnList, newSpendList)
+        addOrders(newEarnList, earnOrderList)
+        addOrders(newSpendList, spendOrderList)
+    }
+
+    private fun addOrders(orders: List<Order>, currentOrders: MutableList<Order>) {
+        if (orders.isNotEmpty()) {
+            //the oldest order is the last one, so we'll go from the last and add the top
+            //we will end with newest order at the top.
+            for (i in orders.indices.reversed()) {
+                val order = orders[i]
+                val index = currentOrders.indexOf(order)
+                if (index == NOT_FOUND) {
+                    //add at top (ui orientation)
+                    addOrder(order = order)
+                } else {
+                    //Update
+                    updateOrder(index = index, order = order)
+                }
+            }
+        }
+    }
+
+    private fun splitByType(list: List<Order>, earnList: MutableList<Order>, spendList: MutableList<Order>) {
```

Comment:
warning! Kotlin show off:
```
    private fun splitByType(list: List<Order>) : Pair<List<Order>, List<Order>>{
        return list.filter { it.status != Status.PENDING }
                .partition { isEarn(it) }
    }
```
and then in `syncNewOrders`:
```
   private fun syncNewOrders(newOrdersListObj: OrderList) {
        val (newEarnList,newSpendList ) = splitByType(newOrdersListObj.orders)
        addOrders(newEarnList, earnOrderList)
        addOrders(newSpendList, spendOrderList)
    }
```

Reply:
OMG

![](https://media.giphy.com/media/fnpxUJaIeJSLRgTEDE/giphy.gif)